uniffi_macros/src/error.rs

```diff
@@ -84,27 +87,49 @@ pub(crate) fn flat_error_ffi_converter_impl(
                     }
                 }
             });
-            let write_impl = quote! {
+            quote! {
                 let error_msg = ::std::string::ToString::to_string(&obj);
                 match obj { #(#write_match_arms)* }
-            };
-
-            write_impl
+            }
         }
         None => quote! { ::std::unimplemented!() },
     };
 
+    let try_read_impl = if implement_try_read {
+        match variants {
+            Some(variants) => {
+                let match_arms = variants.iter().enumerate().map(|(i, v)| {
+                    let v_ident = &v.ident;
+                    let idx = Index::from(i + 1);
+
+                    quote! {
+                        #idx => Self::#v_ident,
+                    }
+                });
+                quote! {
+                    Ok(match ::uniffi::deps::bytes::Buf::get_i32(buf) {
+                        #(#match_arms)*
+                        v => ::uniffi::deps::anyhow::bail!("Invalid #ident enum value: {}", v),
+                    })
+                }
+            }
+            None => quote! { ::std::unimplemented!() },
+        }
+    } else {
+        quote! { ::std::panic!("try_read not supported for flat errors") }
+    };
+
     quote! {
         #[automatically_derived]
-        unsafe impl<T> ::uniffi::FfiConverter<T> for #ident {
-            ::uniffi::ffi_converter_rust_buffer_lift_and_lower!(T);
+        unsafe impl ::uniffi::FfiConverter<crate::UniFfiTag> for #ident {
+            ::uniffi::ffi_converter_rust_buffer_lift_and_lower!(crate::UniFfiTag);
```

Comment:
This changes the code to only implement `FfiConverter<crate::UniFffiTag>`, meaning other crates don't get an automatic implementation.  I made the change because:

- I want to move the interface `FfiConverter` impls to the generate code, but that can't be a blanket impl since `Arc<T>` is not a local type
- The blanket impl doesn't work for remote types and this is currently a requirement for the UDL-based code
- I think we can still make sharing types between crates ergonomic.  What if the requirement was that users needed to write `uniffi_use!(external_crate::ExternalType)`?  That seems reasonable to me.  I hope to have another PR, based on top of this, that implements this.

Reply:
I'm still against changing the default. How about having UDL codegen communicate to the proc-macro that it wants this sort of impl? Something like

```rust
#[derive(uniffi::Enum)]
#[uniffi(local_impl)]
enum Foo { ... }
```

Reply:
Yes, we can do that.